p2p/src/network/identify/stream/p2p_network_identify_stream_reducer.rs

```diff
@@ -2,32 +2,70 @@ use super::{
     P2pNetworkIdentifyStreamAction, P2pNetworkIdentifyStreamKind, P2pNetworkIdentifyStreamState,
 };
 use crate::{
-    network::identify::{pb::Identify, P2pNetworkIdentify},
-    P2pLimits, P2pNetworkStreamProtobufError,
+    identify::P2pIdentifyAction,
+    network::identify::{
+        pb::Identify, stream::P2pNetworkIdentifyStreamError,
+        stream_effectful::P2pNetworkIdentifyStreamEffectfulAction, P2pNetworkIdentify,
+        P2pNetworkIdentifyState,
+    },
+    ConnectionAddr, Data, P2pLimits, P2pNetworkConnectionError, P2pNetworkSchedulerAction,
+    P2pNetworkStreamProtobufError, P2pNetworkYamuxAction, PeerId, YamuxFlags,
 };
-use openmina_core::bug_condition;
+use openmina_core::{bug_condition, warn, Substate, SubstateAccess};
 use prost::Message;
 use quick_protobuf::BytesReader;
-use redux::ActionWithMeta;
+use redux::{ActionWithMeta, Dispatcher};
 
 impl P2pNetworkIdentifyStreamState {
-    pub fn reducer(
-        &mut self,
+    pub fn reducer<Action, State>(
+        mut state_context: Substate<Action, State, P2pNetworkIdentifyState>,
         action: ActionWithMeta<&P2pNetworkIdentifyStreamAction>,
         limits: &P2pLimits,
-    ) -> Result<(), String> {
-        let (action, _meta) = action.split();
-        match &self {
+    ) -> Result<(), String>
+    where
+        State: SubstateAccess<P2pNetworkIdentifyState>,
+        Action: crate::P2pActionTrait<State>,
+    {
+        let (action, meta) = action.split();
+        let substate = state_context.get_substate_mut()?;
+        let stream_state = match action {
+            P2pNetworkIdentifyStreamAction::New {
+                peer_id, stream_id, ..
+            } => substate
+                .create_identify_stream_state(peer_id, stream_id)
+                .map_err(|stream| {
+                    format!("Identify stream already exists for action {action:?}: {stream:?}")
+                })?,
+            P2pNetworkIdentifyStreamAction::Prune {
+                peer_id, stream_id, ..
+            } => {
+                return substate
```

Comment:
@0xMimir one kind of subtle difference between the new-style reducers and the old is that short-circuiting the reducer will not run the code that used to run in the effect handler before. Not about this particular case here just a general comment for you to keep in mind.